src/Compatibility/Microsoft.DotNet.PackageValidation/ApiCompatRunner.cs

```diff
@@ -0,0 +1,94 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.DotNet.ApiCompatibility;
+using Microsoft.DotNet.ApiCompatibility.Abstractions;
+using NuGet.Common;
+
+namespace Microsoft.DotNet.PackageValidation
+{
+    /// <summary>
+    /// Runs ApiCompat over different assembly tuples.
+    /// </summary>
+    public class ApiCompatRunner
+    {
+        private List<(string leftAssemblyPackagePath, string leftAssemblyRelativePath, string rightAssemblyPackagePath, string rightAssemblyRelativePath, string assemblyName, string compatibilityReason, string header)> _queue = new();
+        private ILogger _log;
+        private Checker _checker;
+
+        public ApiCompatRunner(string noWarn, (string, string)[] ignoredDifferences, ILogger log)
+        {
+            _log = log;
+            _checker = new Checker(noWarn, ignoredDifferences, null);
+        }
+
+        /// <summary>
+        /// Runs the api compat for the tuples in the queue.
+        /// </summary>
+        public void RunApiCompat()
+        {
+            foreach (var apicompatTuples in _queue.Distinct())
+            {
+                // TODO: Add Assembly version check.
+                // TODO: Add optimisations tuples.
+                // TODO: Run it Asynchronously.
+                using (Stream leftAssemblyStream = GetFileStreamFromPackage(apicompatTuples.leftAssemblyPackagePath, apicompatTuples.leftAssemblyRelativePath))
+                using (Stream rightAssemblyStream = GetFileStreamFromPackage(apicompatTuples.rightAssemblyPackagePath, apicompatTuples.rightAssemblyRelativePath))
+                {
+                    IAssemblySymbol leftSymbols = new AssemblySymbolLoader().LoadAssembly(apicompatTuples.assemblyName, leftAssemblyStream);
+                    IAssemblySymbol rightSymbols = new AssemblySymbolLoader().LoadAssembly(apicompatTuples.assemblyName, rightAssemblyStream);
+                    ApiComparer differ = new();
+
+                    IEnumerable<CompatDifference> differences = differ.GetDifferences(leftSymbols, rightSymbols).Where(t => !_checker.Contain(t.DiagnosticId, t.ReferenceId));
```

Comment:
Why are you checking these again? Instead flow noWarn and ignoredDifferences to the `differ` as properties and that should filter out the differences for you.